File: expo-app/src/lib/underwriting/underwritingEngine.ts
/**
 * Single entry point: calculate(inputs) → outputs.
 * All formulas in domain modules; deterministic.
 * Matches PropFolio UnderwritingEngine.
 */

import type { UnderwritingInputs, UnderwritingOutputs } from './types';
import * as Income from './incomeFlow';
import * as Debt from './debtAndCashFlow';
import * as Return from './returnMultiplier';
import * as Unit from './unitAndOccupancy';

export function calculate(inputs: UnderwritingInputs): UnderwritingOutputs {
  const gsr = Income.grossScheduledRentAnnual(
    inputs.monthlyRent ?? null,
    inputs.grossScheduledRentAnnual ?? null
  );
  const vacancyAdjusted = Income.vacancyAdjustedGrossIncome(
    gsr,
    inputs.vacancyPercent ?? null
  );
  const otherIncome = Income.otherIncomeAnnual(inputs.otherIncomeAnnual);
  const egi = Income.effectiveGrossIncome(
    vacancyAdjusted,
    inputs.otherIncomeAnnual ?? null
  );
  const oe = Income.operatingExpensesAnnual(inputs.operatingExpensesAnnual ?? null);
  const noi = Income.noi(egi, oe);

  const ads = Debt.annualDebtService(
    inputs.annualDebtService ?? null,
    inputs.loanAmount ?? null,
    inputs.interestRateAnnual ?? null,
    inputs.termYears ?? null
  );

  const monthlyCf = Debt.monthlyCashFlow(noi, ads);
  const annualCf = Debt.annualCashFlow(noi, ads);
  const dscr = Debt.dscr(noi, ads);

  const capRate = Return.capRate(noi, inputs.purchasePrice ?? null);
  const coc = Return.cashOnCashReturn(
    annualCf,
    inputs.purchasePrice ?? null,
    inputs.loanAmount ?? null
  );
  const grm = Return.grm(inputs.purchasePrice ?? null, gsr);
  const expenseRatio = Return.expenseRatio(oe, egi);
  const breakEvenRatio = Return.breakEvenRatio(oe, ads, egi);
  const debtYield = Return.debtYield(noi, inputs.loanAmount ?? null);
  const ltv = Return.ltv(inputs.loanAmount ?? null, inputs.purchasePrice ?? null);

  const pricePerUnit = Unit.pricePerUnit(
    inputs.purchasePrice ?? null,
    inputs.unitCount ?? null
  );
  const pricePerSqFt = Unit.pricePerSquareFoot(
    inputs.purchasePrice ?? null,
    inputs.squareFeet ?? null
  );

  const breakevenOcc = Unit.breakevenOccupancy(
    oe,
    ads,
    gsr,
    inputs.otherIncomeAnnual ?? null
  );
  const paydown5 = Unit.equityPaydown5Year(
    inputs.loanAmount ?? null,
    inputs.interestRateAnnual ?? null,
    inputs.termYears ?? null
  );

  return {
    grossScheduledRentAnnual: gsr ?? undefined,
    vacancyAdjustedGrossIncome: vacancyAdjusted ?? undefined,
    otherIncomeAnnual: otherIncome > 0 ? otherIncome : undefined,
    effectiveGrossIncome: egi ?? undefined,
    operatingExpensesAnnual: oe ?? undefined,
    noi: noi ?? undefined,
    annualDebtService: ads ?? undefined,
    monthlyCashFlow: monthlyCf ?? undefined,
    annualCashFlow: annualCf ?? undefined,
    dscr: dscr ?? undefined,
    capRate: capRate ?? undefined,
    cashOnCashReturn: coc ?? undefined,
    grm: grm ?? undefined,
    expenseRatio: expenseRatio ?? undefined,
    breakEvenRatio: breakEvenRatio ?? undefined,
    debtYield: debtYield ?? undefined,
    ltv: ltv ?? undefined,
    pricePerUnit: pricePerUnit ?? undefined,
    pricePerSquareFoot: pricePerSqFt ?? undefined,
    breakevenOccupancy: breakevenOcc ?? undefined,
    equityPaydown5Year: paydown5 ?? undefined,
  };
}

export type { UnderwritingInputs, UnderwritingOutputs } from './types';

File: PropFolio/Engine/Underwriting/UnderwritingEngine.swift
//
//  UnderwritingEngine.swift
//  PropFolio
//
//  Single entry point: calculate(inputs) → outputs. All formulas in domain calculators.
//

import Foundation

enum UnderwritingEngine {
    /// Compute all underwriting metrics from inputs. Nil outputs where required inputs are missing.
    static func calculate(_ inputs: UnderwritingInputs) -> UnderwritingOutputs {
        // Income flow
        let gsr = IncomeFlowCalculator.grossScheduledRentAnnual(
            monthlyRent: inputs.monthlyRent,
            grossScheduledRentAnnual: inputs.grossScheduledRentAnnual
        )
        let vacancyAdjusted = IncomeFlowCalculator.vacancyAdjustedGrossIncome(
            grossScheduledRentAnnual: gsr,
            vacancyPercent: inputs.vacancyPercent
        )
        let otherIncome = IncomeFlowCalculator.otherIncomeAnnual(inputs.otherIncomeAnnual)
        let egi = IncomeFlowCalculator.effectiveGrossIncome(
            vacancyAdjustedGrossIncome: vacancyAdjusted,
            otherIncomeAnnual: inputs.otherIncomeAnnual
        )
        let oe = IncomeFlowCalculator.operatingExpensesAnnual(inputs.operatingExpensesAnnual)
        let noi = IncomeFlowCalculator.noi(effectiveGrossIncome: egi, operatingExpensesAnnual: oe)

        // Debt service (input or derived)
        let ads = DebtAndCashFlowCalculator.annualDebtService(
            input: inputs.annualDebtService,
            loanAmount: inputs.loanAmount,
            interestRateAnnual: inputs.interestRateAnnual,
            termYears: inputs.termYears
        )

        // Cash flow & DSCR
        let monthlyCf = DebtAndCashFlowCalculator.monthlyCashFlow(noi: noi, annualDebtService: ads)
        let annualCf = DebtAndCashFlowCalculator.annualCashFlow(noi: noi, annualDebtService: ads)
        let dscr = DebtAndCashFlowCalculator.dscr(noi: noi, annualDebtService: ads)

        // Returns & multipliers
        let capRate = ReturnMultiplierCalculator.capRate(noi: noi, purchasePrice: inputs.purchasePrice)
        let coc = ReturnMultiplierCalculator.cashOnCashReturn(
            annualCashFlow: annualCf,
            purchasePrice: inputs.purchasePrice,
            loanAmount: inputs.loanAmount
        )
        let grm = ReturnMultiplierCalculator.grm(purchasePrice: inputs.purchasePrice, grossScheduledRentAnnual: gsr)
        let expenseRatio = ReturnMultiplierCalculator.expenseRatio(operatingExpensesAnnual: oe, effectiveGrossIncome: egi)
        let breakEvenRatio = ReturnMultiplierCalculator.breakEvenRatio(
            operatingExpensesAnnual: oe,
            annualDebtService: ads,
            effectiveGrossIncome: egi
        )
        let debtYield = ReturnMultiplierCalculator.debtYield(noi: noi, loanAmount: inputs.loanAmount)
        let ltv = ReturnMultiplierCalculator.ltv(loanAmount: inputs.loanAmount, purchasePrice: inputs.purchasePrice)

        // Per-unit / per-SF
        let pricePerUnit = UnitAndOccupancyCalculator.pricePerUnit(purchasePrice: inputs.purchasePrice, unitCount: inputs.unitCount)
        let pricePerSqFt = UnitAndOccupancyCalculator.pricePerSquareFoot(purchasePrice: inputs.purchasePrice, squareFeet: inputs.squareFeet)

        // Occupancy & paydown
        let breakevenOcc = UnitAndOccupancyCalculator.breakevenOccupancy(
            operatingExpensesAnnual: oe,
            annualDebtService: ads,
            grossScheduledRentAnnual: gsr,
            otherIncomeAnnual: inputs.otherIncomeAnnual
        )
        let paydown5 = UnitAndOccupancyCalculator.equityPaydown5Year(
            loanAmount: inputs.loanAmount,
            interestRateAnnual: inputs.interestRateAnnual,
            termYears: inputs.termYears
        )

        return UnderwritingOutputs(
            grossScheduledRentAnnual: gsr,
            vacancyAdjustedGrossIncome: vacancyAdjusted,
            otherIncomeAnnual: otherIncome > 0 ? otherIncome : nil,
            effectiveGrossIncome: egi,
            operatingExpensesAnnual: oe,
            noi: noi,
            annualDebtService: ads,
            monthlyCashFlow: monthlyCf,
            annualCashFlow: annualCf,
            dscr: dscr,
            capRate: capRate,
            cashOnCashReturn: coc,
            grm: grm,
            expenseRatio: expenseRatio,
            breakEvenRatio: breakEvenRatio,
            debtYield: debtYield,
            ltv: ltv,
            pricePerUnit: pricePerUnit,
            pricePerSquareFoot: pricePerSqFt,
            breakevenOccupancy: breakevenOcc,
            equityPaydown5Year: paydown5,
            irrPlaceholder: .requiresExplicitSchedule
        )
    }
}
